File: jerryscript/benchmarks/ud.js
var SCALE_FACTOR = 1;

function fillArray(arr, length, value) {
  for (var i = 0; i < length; i++) {
    arr[i] = value;
  }
  return arr;
}

var a = [];
var b = [];
var x = [];
for (var i = 0; i < 20; i++) {
  b[i] = 0;
  x[i] = 0;
  a[i] = [];
  fillArray(a[i], 20, 0);
}

var chkerr = 0;

function ludcmp(nmax, n) {
  var y = [];
  fillArray(y, 100, 0);

  for (var i = 0; i < n; i++) {
    for (var j = i + 1; j <= n; j++) {
      var w = a[j][i];
      if (i !== 0) {
        for (var k = 0; k < i; k++) {
          w = w - a[j][k] * a[k][i];
        }
      }
      a[j][i] = Math.floor(w / a[i][i]);
    }
    for (var j = i + 1; j <= n; j++) {
      var w = a[i + 1][j];
      for (var k = 0; k <= i; k++) {
        w = w - a[i + 1][k] * a[k][j];
      }
      a[i + 1][j] = w;
    }
  }

  y[0] = b[0];
  for (var i = 1; i <= n; i++) {
    var w = b[i];
    for (var j = 0; j < i; j++) {
      w = w - a[i][j] * y[j];
    }
    y[i] = w;
  }

  x[n] = Math.floor(y[n] / a[n][n]);
  for (var i = n - 1; i >= 0; i--) {
    var w = y[i];
    for (var j = i + 1; j <= n; j++) {
      w = w - a[i][j] * x[j];
    }
    x[i] = Math.floor(w / a[i][i]);
  }

  return 0;
}

function initialise_benchmark() {}

function benchmark_body(lsf) {
  for (var iter = 0; iter < lsf; iter++) {
    var nmax = 20;
    var n = 5;
    for (var i = 0; i <= n; i++) {
      var w = 0;
      for (var j = 0; j <= n; j++) {
        a[i][j] = (i + 1) + (j + 1);
        if (i === j) {
          a[i][j] = a[i][j] * 2;
        }
        w = w + a[i][j];
      }
      b[i] = w;
    }
    chkerr = ludcmp(nmax, n);
  }
  return chkerr;
}

function verify_benchmark(res) {
  var x_ref = [0,0,1,1,1,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0];
  if (res !== 0) {
    return false;
  }
  for (var i = 0; i < 20; i++) {
    if (x[i] !== x_ref[i]) {
      return false;
    }
  }
  return true;
}

function benchmark() {
  initialise_benchmark();
  var result = benchmark_body(SCALE_FACTOR);
  return verify_benchmark(result);
}

benchmark();

File: micropython/benchmarks/ud.py
# import micropython

SCALE_FACTOR = 1
a = [[0] * 20 for _ in range(20)]
b = [0] * 20
x = [0] * 20

chkerr = 0


def ludcmp(n: int):
    global a, b, x

    y = [0] * 100

    for i in range(n):
        for j in range(i + 1, n + 1):
            w = a[j][i]
            if i != 0:
                for k in range(i):
                    w -= a[j][k] * a[k][i]

            a[j][i] = w // a[i][i]

        for j in range(i + 1, n + 1):
            w = a[i + 1][j]
            for k in range(i + 1):
                w -= a[i + 1][k] * a[k][j]
            a[i + 1][j] = w

    y[0] = b[0]
    for i in range(1, n + 1):
        w = b[i]
        for j in range(i):
            w -= a[i][j] * y[j]
        y[i] = w

    x[n] = y[n] // a[n][n]
    for i in range(n - 1, -1, -1):
        w = y[i]
        for j in range(i + 1, n + 1):
            w -= a[i][j] * x[j]
        x[i] = w // a[i][i]

    return 0


def verify_benchmark(res: int):
    global x
    x_ref = [0, 0, 1, 1, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

    return x == x_ref and res == 0


def initialize_benchmark():
    pass


def benchmark_body():
    global a, b, chkerr, SCALE_FACTOR
    sf = SCALE_FACTOR

    for _ in range(sf):
        n = 5

        for i in range(n + 1):
            w = 0
            for j in range(n + 1):
                a[i][j] = (i + 1) + (j + 1)
                if i == j:
                    a[i][j] *= 2
                w += a[i][j]
            b[i] = w

        chkerr = ludcmp(n)

        # print("")
        # micropython.mem_info()
    return chkerr


def benchmark():
    initialize_benchmark()
    res = benchmark_body()
    return verify_benchmark(res)
